### PR TITLE
W2 - D1 - thanhhoang - Android Recyclerview

### PR DESCRIPTION
**Pull request by at-thanhhoang**
task #19

### CODE REVIEW

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    @Override
+    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+
+        sArrayImageAvatar = getResources().obtainTypedArray(R.array.imageAvatar);
+        sArrayUsername = getResources().getStringArray(R.array.username_array);
+        sArrayImageFood = getResources().obtainTypedArray(R.array.imageFood);
+        sArrayDescription = new String[]{
```
Comment (at-dev-support-01) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field asiantech.internship.summer.exercise_recycler_view.TimelineFragment.sArrayDescription from instance method asiantech.internship.summer.exercise_recycler_view.TimelineFragment.onActivityCreated(Bundle)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
+        return inflater.inflate(R.layout.fragment_timeline, container, false);
+    }
+
+    @Override
+    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+
+        sArrayImageAvatar = getResources().obtainTypedArray(R.array.imageAvatar);
```
Comment (at-dev-support-01) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field asiantech.internship.summer.exercise_recycler_view.TimelineFragment.sArrayImageAvatar from instance method asiantech.internship.summer.exercise_recycler_view.TimelineFragment.onActivityCreated(Bundle)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        sArrayUsername = getResources().getStringArray(R.array.username_array);
+        sArrayImageFood = getResources().obtainTypedArray(R.array.imageFood);
```
Comment (at-dev-support-01) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field asiantech.internship.summer.exercise_recycler_view.TimelineFragment.sArrayImageFood from instance method asiantech.internship.summer.exercise_recycler_view.TimelineFragment.onActivityCreated(Bundle)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
+        return inflater.inflate(R.layout.fragment_timeline, container, false);
+    }
+
+    @Override
+    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+
+        sArrayImageAvatar = getResources().obtainTypedArray(R.array.imageAvatar);
+        sArrayUsername = getResources().getStringArray(R.array.username_array);
```
Comment (at-dev-support-01) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field asiantech.internship.summer.exercise_recycler_view.TimelineFragment.sArrayUsername from instance method asiantech.internship.summer.exercise_recycler_view.TimelineFragment.onActivityCreated(Bundle)
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/model/TimelineItem.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    public int getImagePost() {
+        return mImagePost;
+    }
+
+    public String getDescriptionFood() {
+        return mDescriptionFood;
+    }
+
+    public boolean getStateLikes() {
```
Comment (at-dev-support-01) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/model/TimelineItem.java]: WARNING: [BooleanGetMethodName] A getX() method which returns a boolean should be named isX()
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#BooleanGetMethodName

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        final Random rand = new Random();
+        final int rndInt = rand.nextInt(mArrayImageFood.length());
+        return mArrayImageFood.getResourceId(rndInt, 0);
+    }
+
+    private String getRandomDescription() {
+        return mArrayDescription[new Random().nextInt(mArrayDescription.length)];
+    }
+
+    public ArrayList<TimelineItem> createTimeLineList() {
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: Em dùng `List` thay vì dùng `ArrayList`

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
+        return inflater.inflate(R.layout.fragment_timeline, container, false);
+    }
+
+    @Override
+    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+
+        mArrayImageAvatar = getResources().obtainTypedArray(R.array.imageAvatar);
+        mArrayUsername = getResources().getStringArray(R.array.username_array);
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: Tên biến này em đặt không đúng convention `username_array`

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    private String[] mArrayDescription = null;
+
+    private ProgressBar mProgressBarLoad;
+    private SwipeRefreshLayout mSwipeRefreshLayout;
+    private TimelineAdapter mAdapterTimeline;
+
+    private List<TimelineItem> mTimelineList;
+
+    private boolean mIsLoading = true;
+    private int mLastVisibleItem, mTotalItemCount, mVisibleItemCount;
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: Mỗi dòng em tạo một biến thôi

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/adapter/TimelineAdapter.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        holder.mImgPostFood.setImageResource(timelineItem.getImagePost());
+        holder.mTvDescriptionFood.setText(Html.fromHtml(timelineItem.getDescriptionFood()), TextView.BufferType.SPANNABLE);
+
+        if (timelineItem.isStateLikes()) {
+            setStateLike(holder.mImgLike, holder.mTvCountLike);
+        } else {
+            setStateDisLike(holder.mImgLike, holder.mTvCountLike);
+        }
+
+        holder.mImgLike.setOnClickListener(view -> {
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/adapter/TimelineAdapter.java]: Sự kiện này em nên set ở trong class `ViewHolder`

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/adapter/TimelineAdapter.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public int getItemCount() {
+        return mTimelineList.size();
+    }
+
+    class TimelineViewHolder extends ViewHolder {
+
+        private CircleImageView mImgAvatar;
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/adapter/TimelineAdapter.java]: Những biến này em nên thêm `final` vào

File: app/src/at-thanhhoang/res/layout/activity_timeline.xml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+
+    <android.support.v7.widget.Toolbar
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/res/layout/activity_timeline.xml]: Em đổi lại như sau
```xml
<android.support.v7.widget.Toolbar
    android:id="@+id/toolbarRecyclerView"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:background="@color/colorPrimary" />
```

File: app/src/at-thanhhoang/res/layout/custom_item_timeline.xml [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginLeft="20dp"
+        android:layout_marginRight="20dp"
+        android:layout_marginTop="10dp"
+        android:singleLine="false"
+        android:textSize="20sp" />
+
+    <View
+        android:layout_width="fill_parent"
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/res/layout/custom_item_timeline.xml]: Em dùng thuộc tính `match_parent`

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    private void fetchData() {
+        mSwipeRefreshLayout.setRefreshing(true);
+        new Handler().postDelayed(() -> {
+            mTimelineList.clear();
+            mTimelineList.addAll(createTimeLineList());
+            mAdapterTimeline.notifyDataSetChanged();
+            mSwipeRefreshLayout.setRefreshing(false);
+        }, 2000);
```
Comment (at-thangtruong) [inline app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineFragment.java]: Em tạo ra một biến như thế này rồi dùng
```java
private static final int TIME_DELAY = 2000`
```

File: app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineActivity.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        FragmentTransaction ft = fragmentManager.beginTransaction();
+        ft.replace(R.id.flContainer, timelineFragment);
+        ft.addToBackStack(null);
+        ft.commit();
+    }
+
+    @Override
+    public void onBackPressed() {
+        super.onBackPressed();
+        if (getFragmentManager().getBackStackEntryCount() > 0) {
```
[the remark this answers is not in the document]
Comment (at-thanhhoang) [reply app/src/at-thanhhoang/java/asiantech/internship/summer/exercise_recycler_view/TimelineActivity.java]: if (getFragmentManager().getBackStackEntryCount() > 0) nó cũng quay lại activity trước a.

**Pull request merged**